File: src/portfolio/generate_website.py
from pathlib import Path
import re
import os

from rst_utils import title, subtitle, card, card_carousel
from generate_resume import create_docx, docx_to_pdf
from data import Portfolio, str_date, Margins, DocxConfig

ROOT = Path(__file__).parent.parent.parent.absolute()

f = open(ROOT / "docs" / "source" / "index.rst", "w")
title(f, "Vikram Rangarajan - Portfolio")


with open(ROOT / "src" / "portfolio" / "portfolio.json") as portfolio_file:
    por = Portfolio.model_validate_json(portfolio_file.read())


# Education Section
subtitle(f, "Education")
education = por.education
if education is not None:
    for edu in education:
        if edu.current is False:
            continue
        time = str_date(edu.start)
        if edu.end is None and edu.expected is not None:
            time += " to Present, Expected " + str_date(edu.expected)
        elif edu.end is not None:
            time += " to " + str_date(edu.end)
        card(
            f,
            f"{edu.name} - {time}",
            f"Degree: {edu.degree}\nMajor: {edu.major}\nMinor: {edu.minor}\nGPA: {edu.GPA}",
        )

# Experience Section
subtitle(f, "Experience & Projects")
experiences = por.experience
if experiences is not None:
    for exp in experiences:
        if exp.current is False:
            continue
        end = str_date(exp.end) if exp.end is not None else "Present"
        body_text = ""
        if exp.title is not None:
            body_text += exp.title
        if exp.description is not None:
            body_text += "\n" + "\n".join(exp.description)
        card(
            f,
            f"{exp.company_name} - {str_date(exp.start)} to {end}",
            body_text,
            exp.title_link,
        )

# Skills Section
skills = por.skills
if skills is not None:
    subtitle(f, "Technical Skills")
    for name in skills.keys():
        skill_group = skills[name]
        name: str = re.sub("[^0-9a-zA-Z]+", " ", name)
        f.write(f"{name.capitalize()}\n\n")
        card_carousel(
            f, skill_group, [None] * len(skill_group), [None] * len(skill_group), 2
        )

# Awards Section
awards = por.awards
if awards is not None:
    subtitle(f, "Awards & Certifications")
    for awd in awards:
        if awd.current is False:
            continue
        awd_date = awd.date if isinstance(awd.date, str) else str_date(awd.date)
        card(
            f,
            f"{awd.name} - {awd_date}",
            None,
            awd.link,
        )


# Create Publications tab
if por.publications is not None:
    f.write("\n.. toctree::\n\t:hidden:\n\n\tpublications")
    with open(ROOT / "docs" / "source" / "publications.rst", "w") as pub_file:
        title(pub_file, "Publications")
        for pub in por.publications:
            authors_str = ", ".join(pub.authors)
            authors_str = authors_str.replace(por.info.name, f"**{por.info.name}**")
            if pub.journal is not None:
                authors_str += f"\nIn {pub.journal}"
            if "published" not in pub.status.lower():
                authors_str += f"\n{pub.status}"
            if pub.date is not None:
                authors_str += str_date(pub.date)
            card(pub_file, pub.title, authors_str, pub.link)


f.write("\n.. toctree::\n\t:hidden:\n\n\tresumes")

# Create Resumes tab
with open(ROOT / "docs" / "source" / "resumes.rst", "w") as res_file:
    title(res_file, "Resume Downloads")
    res_file.write("\n:download:`PDF <_static/resume.pdf>`\n")
    res_file.write("\n:download:`DOCX <_static/resume.docx>`\n\n")
    res_file.write(".. pdf-include:: _static/resume.pdf#view=Fit")

f.close()

# Generate Resume Documents
if not os.path.exists(ROOT / "docs" / "source" / "_static"):
    os.mkdir(ROOT / "docs" / "source" / "_static")
docx_config = DocxConfig(
    margins=Margins(top=0.75, bottom=0.75, left=0.75, right=0.75),
    title_font_size=20,
    heading_font_size=14,
    text_font_size=11,
)
create_docx(ROOT / "docs" / "source" / "_static" / "resume.docx", docx_config)
docx_to_pdf(ROOT / "docs" / "source" / "_static" / "resume.docx")

File: portfolio/src/website/generate_website.rs
use crate::data::*;
use crate::website::rst_utils::*;
use std::fs::{canonicalize, write};
use std::io::Write;
use std::path::Path;
use std::process::Command;

pub fn add_education(mut s: String, p: &Portfolio) -> String {
    s.add_subtitle("Education".into());
    for edu in &p.education {
        if !edu.current {
            continue;
        }
        let mut time = edu.duration.start.str_date();
        match &edu.duration.end {
            EndingDate::End(end) => {
                time.push_str(" to ");
                time.push_str(&end.str_date());
            }
            EndingDate::Current {
                expected: Some(exp),
            } => {
                time.push_str(" to Present, Expected ");
                time.push_str(&exp.str_date());
            }
            EndingDate::Current { expected: None } => {}
        }

        let mut body_str = String::new();

        if let Some(degree) = &edu.degree {
            body_str.push_str(format!("Degree: {degree}\n").as_str())
        }
        if let Some(major) = &edu.major {
            body_str.push_str(format!("Major: {major}\n").as_str())
        }
        if let Some(minor) = &edu.minor {
            body_str.push_str(format!("Minor: {minor}\n").as_str())
        }
        if let Some(courses) = &edu.coursework {
            body_str.push_str(format!("Relevant Coursework: {}\n", courses.join(", ")).as_str());
        }
        body_str.push_str(format!("GPA: {:.1}", edu.gpa).as_str());
        s.add_card(
            Some(format!("{} - {time}", edu.name)),
            Some(body_str),
            None,
            0,
        );
    }
    s
}

pub fn add_experiences(mut s: String, p: &Portfolio) -> String {
    s.add_subtitle("Experience & Projects".into());
    for exp in &p.experience {
        if !exp.current {
            continue;
        }
        let mut time = exp.duration.start.str_date();
        match &exp.duration.end {
            EndingDate::End(end) => {
                time.push_str(" to ");
                time.push_str(&end.str_date());
            }
            EndingDate::Current {
                expected: Some(exp),
            } => {
                time.push_str(" to Present, Expected ");
                time.push_str(&exp.str_date());
            }
            EndingDate::Current { expected: None } => {}
        }

        let mut body_str = String::new();

        if let Some(title) = &exp.title {
            body_str.push_str(title)
        }
        body_str.push_str(format!("\n{}", exp.description.join("\n")).as_str());

        let end_str = match &exp.duration.end {
            EndingDate::End(end) => end.str_date(),
            EndingDate::Current { expected: _ } => "Present".into(),
        };
        s.add_card(
            Some(format!(
                "{} - {} to {}",
                exp.company_name,
                exp.duration.start.str_date(),
                end_str
            )),
            Some(body_str),
            exp.title_link.clone(),
            0,
        );
    }
    s
}

pub fn add_skills(mut s: String, p: &Portfolio) -> String {
    s.add_subtitle("Technical Skills".into());
    for SkillGroup { group_name, skills } in &p.skills {
        s.push_str(format!("{group_name}\n\n").as_str());
        s.add_card_carousel(
            skills.iter().map(|x| Some(x.to_string())).collect(),
            vec![None; skills.len()],
            vec![None; skills.len()],
            2,
        );
    }
    s
}

pub fn add_awards(mut s: String, p: &Portfolio) -> String {
    s.add_subtitle("Awards & Certifications".into());
    for awd in &p.awards {
        if !awd.current {
            continue;
        }
        s.add_card(
            Some(format!("{} - {}", awd.name, awd.date.str_date(),)),
            None,
            awd.link.clone(),
            0,
        );
    }
    s
}

pub fn create_index_file(p: &Portfolio) {
    let mut s = String::with_capacity(10000);
    s.add_title("Portfolio".into());

    s = add_education(s, p);

    s = add_experiences(s, p);

    s = add_skills(s, p);

    s = add_awards(s, p);

    let binding = canonicalize(Path::new(".")).unwrap();
    let root = binding.parent().unwrap();
    let index_filepath = root.join("docs").join("source").join("portfolio.rst");
    write(index_filepath, s).expect("Error Writing File");
}

pub fn create_publications_file(p: &Portfolio) {
    let mut s = String::with_capacity(10000);
    s.add_title("Publications".into());

    for publication in &p.publications {
        let mut authors_str = publication.authors.join(", ");
        authors_str = authors_str.replace(&p.info.name, format!("**{}**", p.info.name).as_str());
        if let Some(journal) = &publication.journal {
            authors_str.push_str(format!("\nIn {}\n", journal).as_str());
        }
        if !publication.status.to_lowercase().contains("published") {
            authors_str.push_str(format!("\n{}", publication.status).as_str());
        }
        if let Some(date) = &publication.date {
            authors_str.push_str(date.str_date().as_str());
        }
        s.add_card(
            Some(publication.title.clone()),
            Some(authors_str),
            publication.link.clone(),
            0,
        );
    }
    let binding = canonicalize(Path::new(".")).unwrap();
    let root = binding.parent().unwrap();
    let index_filepath = root.join("docs").join("source").join("publications.rst");
    write(index_filepath, s).expect("Error Writing File");
}

pub fn create_resumes_file() {
    let mut s = String::with_capacity(10000);
    s.add_title("Resume Downloads".into())
        .push_str("\n:download:`PDF <_static/resume.pdf>`\n");
    s.push_str("\n:download:`DOCX <_static/resume.docx>`\n\n");

    // Add the pdf previewer as an embedded iframe
    s.push_str(".. raw:: html\n\n\t<iframe id=\"pdf-viewer\" ");
    s.push_str("src=\"_static/resume.pdf\" ");
    s.push_str("type=\"application/pdf\" ");
    s.push_str(
        "style=\"height: 60vh; width: 100%\" allowfullscreen scrolling=\"auto\"></iframe>\n",
    );

    let binding = canonicalize(Path::new(".")).unwrap();
    let root = binding.parent().unwrap();
    let index_filepath = root.join("docs").join("source").join("resumes.rst");
    write(index_filepath, s).expect("Error Writing File");
}

pub fn generate_website() {
    let p = load_portfolio();
    create_index_file(&p);
    create_publications_file(&p);
    create_resumes_file();
}

pub fn generate_html() {
    let output = Command::new("make")
        .args(["-C", "../docs", "html"])
        .output()
        .expect("Failed to generate html");
    let _ = std::io::stdout().write_all(&output.stdout);
    let _ = std::io::stderr().write_all(&output.stderr);
}
